operators/pkg/instmetrics/metrics_scraper.go

```diff
@@ -0,0 +1,195 @@
+type CustomMetrics struct {
+	CPUPerc   float32 `json:"cpu"`
+	MemBytes  uint64  `json:"mem"`
+	DiskBytes uint64  `json:"disk"`
+}
+
+func (c CustomMetrics) String() string {
+	return fmt.Sprintf("[CPU] %v \t[MEM] %v \t[DISK] %v", c.CPUPerc, c.MemBytes, c.DiskBytes)
+}
+
+// MetricsScraper interface.
+type MetricsScraper struct {
+	Log           logr.Logger
+	UpdatePeriod  time.Duration
+	RuntimeClient *RemoteRuntimeServiceClient
+	// Containers ignored while retrieving metrics.
+	ignoredContainerNames []string
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=ContainerStats>
+	oldStats map[string]*runtimeapi.ContainerStats
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=CustomMetrics>
+	cachedMetrics      map[string]*CustomMetrics
+	cachedMetricsMutex sync.RWMutex
+}
+
+// Start scraping metrics.
+func (ms *MetricsScraper) Start(ctx context.Context) {
+	ms.ignoredContainerNames = []string{
+		forge.XVncName,
+		forge.WebsockifyName,
+		forge.ContentDownloaderName,
+		forge.ContentUploaderName,
+		forge.MyDriveName,
+	}
+	ms.oldStats = make(map[string]*runtimeapi.ContainerStats)
+
+	// Periodically scrape updated metrics for watched containers.
+	ticker := time.NewTicker(ms.UpdatePeriod)
+	ms.Log.Info("MetricsScraper started", "updatePeriod", ms.UpdatePeriod)
+	for range ticker.C {
+		log := ms.Log.WithValues("timeout", ms.UpdatePeriod)
+		ctx, cancel := context.WithTimeout(ctx, ms.UpdatePeriod)
+
+		if err := ms.fillCachedMetrics(ctx); err != nil {
+			log.Error(err, "Error retrieving containerStats")
+		}
+
+		// Operation running the context completed
+		cancel()
+	}
+}
+
+func (ms *MetricsScraper) fillCachedMetrics(ctx context.Context) error {
+	log := ms.Log.WithName("fill-cached-metrics")
+
+	// Update containerIDs list.
+	containerIDs, err := ms.findApplicationContainerIDs(ctx)
+	if err != nil {
+		log.Error(err, "Error updating containerIDs list")
+	}
+
+	for i, containerID := range containerIDs {
+		log = log.WithValues("containerId", containerID)
+
+		filter := &runtimeapi.ContainerStatsFilter{
+			Id: containerIDs[i],
+		}
+		ctx = clctx.LoggerIntoContext(ctx, log.WithName("list-containers-stats"))
+		containerStatsList, err := ms.RuntimeClient.ListContainerStats(ctx, filter)
+		if err != nil {
+			return err
+		}
+		// If no stats are found for containerID, Pod may no longer be running
+		// and still not being removed from containerIDs list.
+		if len(containerStatsList.GetStats()) == 0 {
+			log.V(2).Info("Removing containerId related cached metrics", "ID", containerID)
+			ms.removeCachedMetric(containerID)
+			continue
+		}
+
+		metrics := &CustomMetrics{}
+		containerStats := containerStatsList.GetStats()[0]
+		metrics.MemBytes = containerStats.GetMemory().GetWorkingSetBytes().GetValue()
+		metrics.DiskBytes = containerStats.GetWritableLayer().GetUsedBytes().GetValue()
+
+		podName := containerStats.Attributes.Labels["io.kubernetes.pod.name"]
```

Comment:
AFAIR, we are currently not depending on the kubelet package. I feel it would be better to define a constant (or even keep it as now, as only used here), to avoid introducing a new (large) dependency.

Reply:
Clearly I haven't checked out where that constant were coming from. Of course if it's not already imported it's not really meaningful to get it from another package. In case we won't setup tests too I guess it can be kept here with no problems, until they decide to change it at least but still I think it's unlikely... And probably it would break many other things before this one 🥴